### PR TITLE
Force IR to be always var=call in case of multi-reg returning call nodes plus refactoring changes

### PR DESCRIPTION
**Pull request by sivarv**
This corresponds to part of sub work item 3 in PR #3941 

Summary of code changes:
- Importer always generates var=call for multi-reg call nodes instead of lcl.fld = call in some cases.  This would mean codegen will see only GT_STORE_LCL_VAR of type struct and never GT_STORE_LCL_FLD of type struct.
- Added methods GetABIReturnReg(int idx)  and GetABIReturnRegs() methods to ReturnTypeDesc.
- Codegen uses ReturnTypeDesc  to know the type descriptor of the method that is being jit compiled.
- method genStructReturnRegisters() is deleted as it got replaced with GetABIReturnREg(int idx). method getStructTypeOffset() is renamed using Pascal naming convention moved to Compiler class from CodegenLinear
- Replaced quite a few places that check whether call node has RetBuf arg with the call to member function HasRetBufArg()

One existing issue was uncovered as part of these changes:  In case of multi-reg struct return in var=call, codegen requires that var is not struct promoted (i.e. lvIsMultiRegArgOrRet = true).  But that is not asserted by codegen.  When I did crossgen of mscorlib with such an assert in place, it fired many times. From that i concluded that we are not setting  lvIsMultiRegArgOrRet = true in all possible code paths where var=call could materialize.   I have opened git issue #4276 to further investigate this.  For now the assert is commented out in this PR.

Desktop DDR is green with no asm diffs.

### CODE REVIEW

**Comment by sivarv** [comment]
@dotnet/jit-contrib - Please review this.

File: src/jit/gentree.cpp [hunk 1]
```diff
@@ -13443,3 +13443,100 @@ void ReturnTypeDesc::Initialize(Compiler* comp, CORINFO_CLASS_HANDLE retClsHnd)
...
+//     idx   -   Index of the return register
+//
+// Return Value:
+//     Returns ith return register as per target ABI.
+//
+// Notes:
+//     Right now this is implemented only for x64 Unix
+regNumber ReturnTypeDesc::GetABIReturnReg(unsigned idx)
+{
+    assert(idx > 0);
```
Comment (CarolEidt) [inline src/jit/gentree.cpp]: Why is this 1-based? Most things in the JIT are zero-based, so this seems quite awkward.

File: src/jit/codegenxarch.cpp [hunk 2]
```diff
@@ -1528,43 +1514,28 @@ CodeGen::genStructReturn(GenTreePtr treeNode)
...
+        getEmitter()->emitIns_R_S(ins_Load(type2), emitTypeSize(type2), reg2, lclVar->gtLclNum, 8);
     }
 
-    // Nothing to do if the op1 of the return statement is a GT_CALL. The call already has the return
-    // registers in the proper return registers. 
-    // This assumes that registers never get spilled. There is an Issue 2966 created to track the need 
-    // for handling the GT_CALL case of two register returns and handle it properly for stress modes 
-    // and potential other changes that may break this assumption.
+    // Assumption: multi-reg return value of a GT_CALL node is never spilled.
+    // TODO: support for multi-reg call nodes.
```
Comment (CarolEidt) [inline src/jit/codegenxarch.cpp]: You have removed the issue reference - I'm not sure what the protocol is for putting issue numbers in source, but at a minimum this should be a TODO-Bug. Also, since you're touching this code, you could check that the call is not marked GTF_SPILLED. We will already hit the above noway_assert if a GT_RELOAD has been inserted between the call and the return.

File: src/jit/codegenxarch.cpp [hunk 3]
```diff
@@ -1948,32 +1919,34 @@ CodeGen::genCodeForTreeNode(GenTreePtr treeNode)
...
         }
         break;
 
     case GT_STORE_LCL_VAR:
-        {
-            if (!genStoreRegisterReturnInLclVar(treeNode))
+        {  
+            // var = call, where call returns a multi-reg return value
+            // case is handled separately.
+            if (treeNode->IsMultiRegStoreToLocal())
```
Comment (CarolEidt) [inline src/jit/codegenxarch.cpp]: Since the only place this can happen is var = call, and we don't anticipate other cases for xarch, I would suggest a more meaningful name such as IsMultiRegCallStoreToLocal().

File: src/jit/codegenxarch.cpp [hunk 4]
```diff
@@ -2631,159 +2605,63 @@ CodeGen::genCodeForTreeNode(GenTreePtr treeNode)
...
-//    For System V AMD64 sistems it returns true if this is a struct and storing of the returned
-//    register value is handled. It returns false otherwise.
-//    For all other targets returns false.
-
-bool
-CodeGen::genStoreRegisterReturnInLclVar(GenTreePtr treeNode)
+//    None
+//
+// Note
+//    Right now this is implemented only for x64 Unix.
```
Comment (CarolEidt) [inline src/jit/codegenxarch.cpp]: This note seems to imply that this is a "NYI" for x64 windows, but this will never apply there.

Comment (CarolEidt) [inline src/jit/codegenxarch.cpp]: Although it is stated in the description, you should explicitly state the assumption/pre-condition that the child of the store is a call node.

File: src/jit/codegenxarch.cpp [hunk 5]
```diff
@@ -2631,159 +2605,63 @@ CodeGen::genCodeForTreeNode(GenTreePtr treeNode)
...
+    regNumber reg1 = retTypeDesc->GetABIReturnReg(1);
+    regNumber reg2 = retTypeDesc->GetABIReturnReg(2);
 
-        getStructTypeOffset(structDesc, &type0, &type1, &offset0, &offset1);
-        getStructReturnRegisters(type0, type1, &retReg0, &retReg1);
+    assert(reg1 != REG_NA && reg2 != REG_NA);
 
-        assert(retReg0 != REG_NA && retReg1 != REG_NA);
+    // Assumption: multi-reg return value of a GT_CALL node never gets spilled.
+    // TODO - support for multi-reg GT_CALL nodes.
```
Comment (CarolEidt) [inline src/jit/codegenxarch.cpp]: TODO-BUG

File: src/jit/compiler.cpp [hunk 6]
```diff
@@ -4713,165 +4713,165 @@ unsigned getMethodBodyChecksum(__in_z char *code, int size)
...
-        CORINFO_METHOD_HANDLE methodHnd = info.compMethodHnd;
+int           Compiler::compCompileHelper(CORINFO_MODULE_HANDLE            classPtr,
+                                          COMP_HANDLE                      compHnd,
+                                          CORINFO_METHOD_INFO            * methodInfo,
+                                          void *                         * methodCodePtr,
+                                          ULONG                          * methodCodeSize,
+                                          CORJIT_FLAGS                   * compileFlags,
+                                          CorInfoInstantiationVerification instVerInfo)
+{
+    CORINFO_METHOD_HANDLE methodHnd = info.compMethodHnd;
```
Comment (CarolEidt) [inline src/jit/compiler.cpp]: If you find it necessary to reformat this whole method, you should also bring it up to coding convention standards. Otherwise, I think that you should not change the formatting.

File: src/jit/compiler.h [hunk 7]
```diff
@@ -8817,12 +8817,16 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 
     static HelperCallProperties s_helperCallProperties;
 
-#if defined(FEATURE_UNIX_AMD64_STRUCT_PASSING)
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
```
Comment (CarolEidt) [inline src/jit/compiler.h]: I believe that it is the convention that "FEATURE" tags should always be used as "#if FEATURE_UNIX_AMD64_STRUCT_PASSING. That is, they can be defined as 0 meaning off - that wouldn't work in this case. I realize that you have not changed the semantics, but I'm not sure why you changed it if you are not fixing it.

File: src/jit/gentree.h [hunk 8]
```diff
@@ -2463,10 +2471,34 @@ struct GenTreeCall final : public GenTree
 
     regMaskTP         gtCallRegUsedMask;      // mask of registers used to pass parameters
 
+    // For now Return Type Descriptor is enabled only for x64 unix.
+    // TODO: enable for all arch that support multi-reg returns (i.e. arm64/arm32/x86)
```
Comment (CarolEidt) [inline src/jit/gentree.h]: I believe that coding conventions require all TODOs to be qualified. I think that there are both TODO-X86 and TODO-ARM as well as TODO-Cleanup items (the latter for targets that are currently fully functional but that we want to unify).

File: src/jit/gentree.h [hunk 9]
```diff
@@ -2538,10 +2570,27 @@ struct GenTreeCall final : public GenTree
...
+    //
+    // Note:
+    //     This is implemented only for x64 Unix and yet to be implemented for
+    //      arm64/arm32/x86
+    bool HasMultiRegRetVal() const 
+    { 
 #ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
-    // Returns true if the call is returning a multi-reg struct value
-    bool HasMultiRegRetVal() const { return varTypeIsStruct(gtType) && !HasRetBufArg(); }
+        return varTypeIsStruct(gtType) && !HasRetBufArg();
```
Comment (CarolEidt) [inline src/jit/gentree.h]: There should probably be an assert that this is a GT_CALL.

File: src/jit/importer.cpp [hunk 10]
```diff
@@ -1131,18 +1133,28 @@ GenTreePtr Compiler::impAssignStructPtr(GenTreePtr      dest,
         }
         else
         {
+            // Case of call returning a struct in register(s)
+
```
Comment (CarolEidt) [inline src/jit/importer.cpp]: I would reword this as "Case of a call returning a struct in one or more registers." (Also, include a period as per CC).

File: src/jit/importer.cpp [hunk 11]
```diff
@@ -13676,6 +13688,10 @@ bool Compiler::impReturnInstruction(BasicBlock *block, int prefixFlags, OPCODE &
             // we don't unconditionally overwrite it, it shouldn't matter.
             if (info.compRetNativeType != TYP_STRUCT)
             {
+                // compretNativeType is not TYP_STRUCT.
```
Comment (CarolEidt) [inline src/jit/importer.cpp]: typo: compRetNativeType

File: src/jit/importer.cpp [hunk 12]
```diff
@@ -13801,13 +13821,15 @@ bool Compiler::impReturnInstruction(BasicBlock *block, int prefixFlags, OPCODE &
                 {
                     // Same as !IsHfa but just don't bother with impAssignStructPtr.
 #else // !defined(_TARGET_ARM_)
-                SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR structDesc;
-                eeGetSystemVAmd64PassStructInRegisterDescriptor(retClsHnd, &structDesc);
-                if (structDesc.passedInRegisters)
+                ReturnTypeDesc retTypeDesc;
```
Comment (CarolEidt) [inline src/jit/importer.cpp]: It appears that there is a (pre-existing) indentation problem here - could you fix it since you're changing this code?

File: src/jit/compiler.cpp (hunk 6 above)
Comment (sivarv) [inline src/jit/compiler.cpp]: I didn't touch this method but while creating PR i have noticed there were formatting changes.  That was a bit of surprise for me too.  It could be an accidental change or VS editor might have auto formatted it.

I disagree that one has to also bring a method to CC standards just because they edited the formatting (intentionally or unintentionally).

**Comment by CarolEidt** [comment]
I like the direction but have a number of structure and CC suggestions.

File: src/jit/compiler.cpp (hunk 6 above)
Comment (CarolEidt) [inline src/jit/compiler.cpp]: The problem with massive changes is that they clutter the PR review as well as the source history. When they are mixed with actual code changes it makes it that much harder to review the code, and to understand the code changes later. I don't think it is useful to make reformatting changes unless the code is being brought up to CC standards.

File: src/jit/compiler.h (hunk 7 above)
Comment (sivarv) [inline src/jit/compiler.h]: I think the convention is to use #ifdef if the FEATURE tag is defined in build scripts.  If it is defined in target.h, we are using #if. The reason i have changed it to #ifdef is that throughout Jit sources we are using #if defined(xxx) only if the condition is a compound condition

e.g. #if defined(xx) && defined(yyy)

Whereas #ifdef xxx instead of #if defined(xxx).

File: src/jit/gentree.h (hunk 9 above)
Comment (sivarv) [inline src/jit/gentree.h]: This is a member method of GenTreeCall and hence no need to assert that the gentree node is GT_CALL

File: src/jit/gentree.h (hunk 8 above)
Comment (sivarv) [inline src/jit/gentree.h]: Ok. I have added multiple TODO-arch specific

File: src/jit/importer.cpp (hunk 10 above)
Comment (sivarv) [inline src/jit/importer.cpp]: Done.

File: src/jit/importer.cpp (hunk 12 above)
Comment (sivarv) [inline src/jit/importer.cpp]: There is no indentation problem.  The If condition is different for Arm and x64 Unix.  In case of x64 Unix there are 3 lines of code before the if condition.

File: src/jit/gentree.cpp (hunk 1 above)
Comment (sivarv) [inline src/jit/gentree.cpp]: Personally, I find 1-based is easier to understand. But if all things are zero-based, for consistency i will change this to zero based.

**Comment by sivarv** [comment]
@CarolEidt - I have incorporated review feedback.  Please review this.

**Comment by sivarv** [comment]
Desktop DDR is green with no asm diffs.

File: src/jit/compiler.cpp (hunk 6 above)
Comment (CarolEidt) [inline src/jit/compiler.cpp]: I would still prefer not to see this unrelated reformatting as part of this PR.

File: src/jit/compiler.h (hunk 7 above)
Comment (CarolEidt) [inline src/jit/compiler.h]: I won't object to leaving this as-is, but see https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-jit-coding-conventions.md#141-conditional-compilation

File: src/jit/importer.cpp (hunk 12 above)
Comment (CarolEidt) [inline src/jit/importer.cpp]: Oh, man - that is some really ugly formatting. Hopefully we can clean that up in future.

File: src/jit/compiler.cpp [hunk 13]
```diff
@@ -5815,16 +5819,52 @@ var_types Compiler::getEightByteType(const SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASS
...
-        assert(false && "getEightByteType Invalid classification type.");
+        assert(false && "GetEightByteType Invalid classification type.");
         break;
     }
 
     return eightByteType;
 }
+
+//------------------------------------------------------------------------------------------------------
+// GetStructTypeOffset: Gets the type, size and offset of the eightbytes of a struct for System V systems.
```
Comment (LLITCHEV) [inline src/jit/compiler.cpp]: It looks like the function doesn't return size anymore.

**Comment by LLITCHEV** [comment]
Looks fine, except the minor mention of "size" in the comment to one of the functions.

File: src/jit/compiler.cpp (hunk 13 above)
Comment (sivarv) [inline src/jit/compiler.cpp]: I just moved this routine from codegenlinear.h/codegenxarch.cpp to compiler.h/compiler.cpp verbatim and didn't change function implementation.

Function return is still void, it just that it returns the values through out params, which is indicated in Arguments" section.

**Comment by sivarv** [comment]
I have updated PR to exclude formatting changes in compiler.cpp.

**Comment by CarolEidt** [comment]
LGTM - thanks

**Comment by sivarv** [comment]
@dotnet-bot retest  OSX x64 Checked Build and Test

**Comment by sivarv** [comment]
@dotnet-bot retest Windows_NT x86 Debug Legacy Backend Build and Test

**Comment by sivarv** [comment]
@dotnet-bot retest OSX x64 Checked Build and Test

**Comment by sivarv** [comment]
@dotnet-bot retest Windows_NT x86 Debug Legacy Backend Build and Test

**Comment by sivarv** [comment]
OSX x64 Checked failure is an infra issue.  It passed earlier before I have updated PR to exclude formatting changes to compiler.cpp.  Going ahead with merge.
